pkg/oc/admin/prune/images.go

```diff
@@ -412,6 +417,10 @@ func (o PruneImagesOptions) Run() error {
 		fmt.Fprintln(o.ErrOut, "Dry run enabled - no modifications will be made. Add --confirm to remove images")
 	}
 
+	if o.LeaveRegistryAlone != nil && *o.LeaveRegistryAlone {
+		fmt.Fprintln(o.Out, "Soft prune enabled - only API objects will be removed.  No modifications to the image registry will be made.")
```

Comment:
Please, don't call it soft prune. It doesn't make any sense.

LeaveRegistryAlone makes the kube objects pruner more robust, but it still may remove too much, it doesn't make the pruner safer.

We have pruners for kubernetes objects (and I think we should split it in two smaller parts: for image streams and for images) and for a registry storage. Both of them are pretty dangerous, but both of them must be used by customers in some cases. None of them should be emphasised as safer and more delicate.

Reply:
+1 on the "it is not safer" point @dmage - while that was in the card's description originally, that discussion point came up during planning as the comment was removed from the card description.

I'm fine for not using "soft prune" ... this does IMO tie in with the discussion / struggle with what the name of the parameter should be (see my comments above where I explicitly did not using `--soft-prune`).

I'm open to suggestions on the message wording .... the pattern does not mimic the `Dry run enabled - ..` message above, but how ab out simply removing `Soft prune enabled - ` and start with `Only API objects will be...` ?

On the splitting things up to an option for only deleting image streams and an option for only deleting images, I don't think I can assess the strength of the potentially decreased usability (i.e. more options) vs. the increased safety of further dissecting things.  I would defer to you and the other SMEs here, though think I should wait on @bparees take before committing.

Reply:
I wouldn't mind keeping the soft prune naming when accompanied by a strong warning that the registry will be left with unprunable objects. Prunable only via hard-prune. I like the idea of hard prune being a complement of soft prune. What should we call the regular prune :smile:?

Reply:
re: pruning imagestreams and images separately:
removing images w/o updating imagestreams just creates a bunch of broken content.
removing imagestreams w/o removing images just creates a bunch of hidden/mostly-inaccessible content

so while I appreciate the idea from a perspective of reducing the granularity and potentially making each operation cleaner/safer, I don't think there's a compelling motivation for it right now.  Either way it would be a follow on to this effort if we did it at all, so i think we can table it for now.

For the text, i'm ok w/ just removing the soft prune part.

Reply:
10-4 part of next push thanks

Reply:
> removing images w/o updating imagestreams just creates a bunch of broken content.

I didn't suggest this. This breaks consistency ans it's a wrong way.

> removing imagestreams w/o removing images just creates a bunch of hidden/mostly-inaccessible content

Yep, exactly what we are doing in this PR. Though, these hidden content can be reused later.

Reply:
@dmage when i said "image" i meant the image etcd object, not the registry data.  So given that understanding:

>> removing images w/o updating imagestreams just creates a bunch of broken content.
removing imagestreams w/o removing images just creates a bunch of hidden/mostly-inaccessible content
> I didn't suggest this. This breaks consistency ans it's a wrong way.

isn't that exactly what you're suggestring here (remove image etcd objects and imagestream etcd objects independently):

> We have pruners for kubernetes objects (and I think we should split it in two smaller parts: for image streams and for images) 

and:

>> removing imagestreams w/o removing images just creates a bunch of hidden/mostly-inaccessible content
> Yep, exactly what we are doing in this PR. Though, these hidden content can be reused later.

no, what we're doing in this PR is removing the image+imagestreamtag etcd data but not removing the registry blob data.  

So to summarize, there are 3 types of things that get pruned:
1) registry blob data
2) image etcd definitions
3) imagestreamtag references to image etcd definitions

This PR is going to allow 1 and 2/3 to be done independently.

It sounds like you were suggesting 2+3 should also be possible to perform independently, and I am suggesting that doing 2+3 independently doesn't add value.

Reply:
When you remove image streams and images from etcd, you create a bunch of hidden/mostly-inaccessible content (I mean blobs on the storage).

Removing images from etcd (2) is just a garbage collection routine (remove images that are not referenced in any image stream), in theory it should be absolutely safe and it can be performed autonomously. It doesn't visible for a regular user, only cluster admins care about it.

On the other side, removing tags from image streams and removing image streams from etcd (3) may affect users: it makes some data inaccessible. And, as the integrated registry may be exposed, only the user may remove items from his image streams in a safe way. We cannot guarantee safety of this prune in general, but it still useful for a lot of users.

Therefore, 2 and 3 provide different safety levels. That's why I suggest to split this functionality.

Reply:
ok, i can see the value in being able to prune just the un-referenced image etcd objects.  we should consider it for the future.

Reply:
I would argue that 2 is completely:

1. pruner decides to prune image I since it's not referenced
2. user re-pushes I to his repository
3. pruner deletes I
4. user has imagestreamtag pointing to the deleted I

I don't think we can put the burden of image stream pruning on all the cluster users - even if we do, there will always be irresponsible users building many images without pruning (e.g. online). Admins will want to prune their namespaces as well saying "each user can have up to 3 image revisions per istag, up to 5 is, etc...". With [istag pruning immunity](#17511) in place, users can request it for their imagestream(tag)s and be safe from the dangerous pruning.

Reply:
@miminar that could only happen if the user re-pushes image I w/ the exact same imageid *and* does so at the same time that pruning is running and selected image I for pruning because it was not referenced by an imagestream.  So a risk, but a pretty small one.

Anyway again, these are all discussion points for the larger "how do we rethink pruning" conversation, but independent of the incremental improvement in this PR.